sdk/schemaregistry/azure-schemaregistry-avroserializer/azure/schemaregistry/serializer/avroserializer/aio/_schema_registry_avro_serializer_async.py

```diff
@@ -0,0 +1,173 @@
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+try:
+    from functools import lru_cache
+except ImportError:
+    from backports.functools_lru_cache import lru_cache
+from io import BytesIO
+from typing import Any, Dict, Mapping
+import avro
+
+from .._constants import SCHEMA_ID_START_INDEX, SCHEMA_ID_LENGTH, DATA_START_INDEX
+from .._avro_serializer import AvroObjectSerializer
+
+
+class AvroSerializer(object):
+    """
+    AvroSerializer provides the ability to serialize and deserialize data according
+    to the given avro schema. It would automatically register, get and cache the schema.
+
+    :keyword client: Required. The schema registry client
+     which is used to register schema and retrieve schema from the service.
+    :paramtype client: ~azure.schemaregistry.SchemaRegistryClient
+    :keyword str group_name: Required. Schema group under which schema should be registered.
+    :keyword bool auto_register_schemas: When true, register new schemas passed to serialize.
+     Otherwise, and by default, fail if it has not been pre-registered in the registry.
+
+    """
+
+    def __init__(self, **kwargs):
+        # type: (Any) -> None
+        try:
+            self._schema_group = kwargs.pop("group_name")
+            self._schema_registry_client = kwargs.pop("client") # type: "SchemaRegistryClient"
+        except KeyError as e:
+            raise TypeError("'{}' is a required keyword.".format(e.args[0]))
+        self._avro_serializer = AvroObjectSerializer(codec=kwargs.get("codec"))
+        self._auto_register_schemas = kwargs.get("auto_register_schemas", False)
+        self._auto_register_schema_func = (
+                self._schema_registry_client.register_schema
+                if self._auto_register_schemas
+                else self._schema_registry_client.get_schema_id
+            )
+
+    async def __aenter__(self):
+        # type: () -> SchemaRegistryAvroSerializer
+        await self._schema_registry_client.__aenter__()
+        return self
+
+    async def __aexit__(self, *exc_details):
+        # type: (Any) -> None
+        await self._schema_registry_client.__exit__(*exc_details)
+
+    async def close(self):
+        # type: () -> None
+        """This method is to close the sockets opened by the client.
+        It need not be used when using with a context manager.
+        """
+        await self._schema_registry_client.close()
+
+    #@lru_cache(maxsize=128)
```

Comment:
functools.lru_cache does not work with async. I found this alternative: https://github.com/aio-libs/async_lru , suggested by the Python core team here: https://bugs.python.org/issue35040
can we add this as a dependency?